### PR TITLE
light: correctly handle contexts

### PR DESCRIPTION
**Pull request by cmwaters**
This is the update to master for https://github.com/tendermint/tendermint/pull/6685 which fixes incorrect handling of contexts within the light client.

### CODE REVIEW

**Review by tychoish — APPROVED** [review]

**Review by tychoish — APPROVED** [review]

File: internal/statesync/syncer.go [hunk 1]
```diff
@@ -38,12 +38,16 @@ var (
...
+	// errVerifyFailed is returned by Sync() when app hash or last height
+	// verification fails.
 	errVerifyFailed = errors.New("verification failed")
 	// errTimeout is returned by Sync() when we've waited too long to receive a chunk.
 	errTimeout = errors.New("timed out waiting for chunk")
 	// errNoSnapshots is returned by SyncAny() if no snapshots are found and discovery is disabled.
 	errNoSnapshots = errors.New("no suitable snapshots found")
+	// errDeadlineExceeded is returned by Sync() when the timeout for retrieving
+	// tendermint state or the commit is exceeded
+	errDeadlineExceeded = errors.New("deadline to retrieve state and commit exceeded")
```
Comment (williambanfield) [inline internal/statesync/syncer.go:50]: This already is an error value in the standard library, can we just use that one?

File: internal/statesync/syncer.go [hunk 2]
```diff
@@ -275,10 +283,17 @@ func (s *syncer) Sync(ctx context.Context, snapshot *snapshot, chunks *chunkQueu
 	// Optimistically build new state, so we don't discover any light client failures at the end.
 	state, err := s.stateProvider.State(pctx, snapshot.Height)
 	if err != nil {
+		if err == context.DeadlineExceeded {
+			return sm.State{}, nil, errDeadlineExceeded
```
Comment (williambanfield) [inline internal/statesync/syncer.go:287]: we wrap the error that we return using the [`%w` format directive](https://blog.golang.org/go1.13-errors#TOC_3.3.). I'm not sure there's a need for this extra case. The caller can use `error.Is(err, Context.DeadlineExceeded)`.

File: internal/statesync/syncer.go (hunk 1 above)
Comment (cmwaters) [reply internal/statesync/syncer.go:50]: The reason why I made a separate error is what happens if someone uses a deadline in the parent context. If it gets triggered we have (AFAIK and context is relatively new to me so I don't know too much) no idea whether the deadline was from the parent context or the sub context we use for retrieving the state and commit information. Does that make sense?

Comment (williambanfield) [reply internal/statesync/syncer.go:50]: I see that this used on line 287 and 295 of this file. In either of those cases, the parent or the child could have timed out and we sill return this error. We still can't tell which one timed out when we precolate this error up but perhaps I'm not totally understanding?

Comment (cmwaters) [reply internal/statesync/syncer.go:50]: Ok good catch. I think I've added the adequate clause to distinguish between parent context errors and child context errors

File: internal/statesync/syncer.go [hunk 3]
```diff
@@ -38,12 +38,16 @@ var (
...
+	// errVerifyFailed is returned by Sync() when app hash or last height
+	// verification fails.
 	errVerifyFailed = errors.New("verification failed")
 	// errTimeout is returned by Sync() when we've waited too long to receive a chunk.
 	errTimeout = errors.New("timed out waiting for chunk")
 	// errNoSnapshots is returned by SyncAny() if no snapshots are found and discovery is disabled.
 	errNoSnapshots = errors.New("no suitable snapshots found")
+	// errStateCommitTimeout is returned by Sync() when the timeout for retrieving
+	// tendermint state or the commit is exceeded
+	errStateCommitTimeout = errors.New("timed out trying to retrieve state and commit")
```
Comment (williambanfield) [inline internal/statesync/syncer.go:50]: :)

**Review by williambanfield — APPROVED** [review]

**Pull request merged**